src/EventEmitterInterface.php

```diff
@@ -13,10 +13,38 @@
 
 interface EventEmitterInterface
 {
-    public function on($event, callable $listener);
-    public function once($event, callable $listener);
-    public function removeListener($event, callable $listener);
-    public function removeAllListeners($event = null);
-    public function listeners($event = null);
-    public function emit($event, array $arguments = []);
+    /**
+     * Allows you to subscribe to an event.
+     */
+    public function on(string $event, callable $listener): static;
+
+    /**
+     * Convenience method that adds a listener which is guaranteed to only be
+     * called once.
+     */
+    public function once(string $event, callable $listener): static;
+
+    /**
+     * Remove a specific listener for a specific event.
+     */
+    public function removeListener(string $event, callable $listener): void;
+
+    /**
+     * Remove all listeners for a specific event or all listeners all together.
+     *
+     * This is useful for long-running processes, where you want to remove listeners
+     * in order to allow them to get garbage collected.
+     */
+    public function removeAllListeners(?string $event = null): void;
+
+    /**
+     * Allows you to inspect the listeners attached to an event. Particularly useful
+     * to check if there are any listeners at all.
+     */
+    public function listeners(?string $event = null): array;
```

Comment:
Good call. Looking back now I'm not sure why I had the array shape in the trait, when it could've just been added to the docblock in the interface. :)

Reply:
Added in latest commit.